test/typecheck/main.go

```diff
@@ -0,0 +1,351 @@
+		}
+		a.typeCheck(dir, a.filterFiles(p.Files))
+	}
+}
+
+// filterFiles restricts a list of files to only those that should be built by
+// the current platform. This includes both build suffixes (_windows.go) and build
+// tags ("// +build !linux" at the beginning).
+func (a *analyzer) filterFiles(fs map[string]*ast.File) []*ast.File {
+	files := []*ast.File{}
+	for _, f := range fs {
+		fpath := a.fset.File(f.Pos()).Name()
+		dir, name := filepath.Split(fpath)
+		matches, err := a.ctx.MatchFile(dir, name)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "%sERROR reading %s: %s\n", logPrefix, fpath, err)
+			a.failed = true
+			continue
+		}
+		if matches {
+			files = append(files, f)
+		}
+	}
+	return files
+}
+
+func (a *analyzer) typeCheck(dir string, files []*ast.File) error {
+	info := types.Info{
+		Defs: make(map[*ast.Ident]types.Object),
+		Uses: make(map[*ast.Ident]types.Object),
+	}
+
+	// NOTE: this type check does a *recursive* import, but srcimporter
+	// doesn't do a full type check (ignores function bodies)-- this has
+	// some additional overhead.
+	//
+	// This means that we need to ensure that typeCheck runs on all
+	// code we will be compiling.
+	//
+	// TODO(rmmh): Customize our forked srcimporter to do this better.
+	pkg, err := a.conf.Check(dir, a.fset, files, &info)
+	if err != nil {
+		return err // type error
+	}
+
+	// A significant fraction of vendored code only compiles on Linux,
+	// but it's only imported by code that has build-guards for Linux.
+	// Track vendored code to type-check it in a second pass.
+	for _, imp := range pkg.Imports() {
+		if strings.HasPrefix(imp.Path(), "k8s.io/kubernetes/vendor/") {
+			vendorPath := imp.Path()[len("k8s.io/kubernetes/"):]
+			if *verbose {
+				fmt.Println("recursively checking vendor path:", vendorPath)
+			}
+			a.collect(vendorPath)
+		}
+	}
+
+	if *defuses {
+		for id, obj := range info.Defs {
+			fmt.Printf("%s: %q defines %v\n",
+				a.fset.Position(id.Pos()), id.Name, obj)
+		}
+		for id, obj := range info.Uses {
+			fmt.Printf("%s: %q uses %v\n",
+				a.fset.Position(id.Pos()), id.Name, obj)
+		}
+	}
+
+	return nil
+}
+
+type collector struct {
+	dirs []string
+}
+
+// handlePath walks the filesystem recursively, collecting directories,
+// ignoring third_party and staging directories.
+func (c *collector) handlePath(path string, info os.FileInfo, err error) error {
+	if err != nil {
+		return err
+	}
+	if info.IsDir() {
+		// Ignore hidden directories (.git, .cache, etc)
+		if len(path) > 1 && path[0] == '.' ||
+			// Staging code is symlinked from vendor/k8s.io, and uses import
+			// paths as if it were inside of vendor/. It fails typechecking
+			// inside of staging/, but works when typechecked as part of vendor/.
+			path == "staging" ||
+			// OS-specific vendor code tends to be imported by OS-specific
+			// packages. We recursively typecheck imported vendored packages for
+			// each OS, but don't typecheck everything for every OS.
+			path == "vendor" ||
```

Comment:
Does this give us the coverage we need, based on the above assertion that we will test staging via vendor? This seems like there could possibly be overlooked type checks in staging if they are only imported by certain os-specific packages.

Reply:
There are os-specific packages that don't even bother with build tags (hcsshim)-- I don't think we need to prove that all vendored code works on all platforms, just that all *built* vendored code works for each platform.